File: windows/Tests/CliArgumentTests.cs
using QuadClicker.Core;
using QuadClicker.Models;
using Xunit;

namespace QuadClicker.Tests;

/// <summary>Tests for CLI argument parsing via ClickRateParser (public surface).</summary>
public sealed class CliArgumentTests
{
    // These tests exercise the parser through the same path the CLI uses.

    [Theory]
    [InlineData("100ms",  100)]
    [InlineData("10/s",   100)]
    [InlineData("600/min", 100)]
    public void Rate_AllFormats_ParseCorrectly(string rate, double expectedMs)
    {
        Assert.True(ClickRateParser.TryParse(rate, out var delay, out _));
        Assert.Equal(expectedMs, delay.TotalMilliseconds, precision: 0);
    }

    [Fact]
    public void Rate_Missing_ReturnsError()
    {
        bool ok = ClickRateParser.TryParse("", out _, out string err);
        Assert.False(ok);
        Assert.NotEmpty(err);
    }

    [Fact]
    public void ClickSession_AllFields_RoundTrip()
    {
        var session = new ClickSession(
            ClickRate:          TimeSpan.FromMilliseconds(200),
            Button:             MouseButton.Right,
            ClickType:          ClickType.Double,
            UseCurrentPosition: false,
            X:                  640,
            Y:                  480,
            StopAfterClicks:    100,
            StopAfterSeconds:   30,
            IdleWaitSeconds:    5);

        Assert.Equal(200,             session.ClickRate.TotalMilliseconds);
        Assert.Equal(MouseButton.Right, session.Button);
        Assert.Equal(ClickType.Double, session.ClickType);
        Assert.False(session.UseCurrentPosition);
        Assert.Equal(640, session.X);
        Assert.Equal(480, session.Y);
        Assert.Equal(100, session.StopAfterClicks);
        Assert.Equal(30,  session.StopAfterSeconds);
        Assert.Equal(5,   session.IdleWaitSeconds);
    }

    [Theory]
    [InlineData(MouseButton.Left)]
    [InlineData(MouseButton.Right)]
    [InlineData(MouseButton.Middle)]
    public void MouseButton_AllValues_AreValid(MouseButton btn)
    {
        var s = new ClickSession(TimeSpan.FromMilliseconds(100), btn,
            ClickType.Single, true, 0, 0, 0, 0, 0);
        Assert.Equal(btn, s.Button);
    }
}

File: linux/tests/CliArgumentTests.cpp
#include <QtTest/QtTest>
#include "../src/core/ClickRateParser.h"
#include "../src/models/ClickSession.h"
#include "../src/models/MouseButton.h"
#include "../src/models/ClickType.h"

using namespace QuadClicker;

/// Tests for CLI argument parsing logic.
/// These exercise the same public parsing surface that the CLI uses
/// (ClickRateParser::tryParse + ClickSession construction).
class CliArgumentTests : public QObject {
    Q_OBJECT

private slots:
    // ── Rate format round-trips ────────────────────────────────────────────────

    void rate_100ms_parsesCorrectly()
    {
        std::chrono::milliseconds delay;
        QString err;
        QVERIFY(ClickRateParser::tryParse(QStringLiteral("100ms"), delay, err));
        QCOMPARE(delay.count(), 100LL);
    }

    void rate_10perS_parsesCorrectly()
    {
        std::chrono::milliseconds delay;
        QString err;
        QVERIFY(ClickRateParser::tryParse(QStringLiteral("10/s"), delay, err));
        QCOMPARE(delay.count(), 100LL);
    }

    void rate_600perMin_parsesCorrectly()
    {
        std::chrono::milliseconds delay;
        QString err;
        QVERIFY(ClickRateParser::tryParse(QStringLiteral("600/min"), delay, err));
        QCOMPARE(delay.count(), 100LL);
    }

    void rate_missing_returnsError()
    {
        std::chrono::milliseconds delay;
        QString err;
        QVERIFY(!ClickRateParser::tryParse(QStringLiteral(""), delay, err));
        QVERIFY(!err.isEmpty());
    }

    // ── Session field round-trips ──────────────────────────────────────────────

    void clickSession_allFields_roundTrip()
    {
        ClickSession session(
            std::chrono::milliseconds(200),
            MouseButton::Right,
            ClickType::Double,
            false,
            640, 480,
            100,
            30.0,
            5.0);

        QCOMPARE(session.clickRate.count(),   200LL);
        QCOMPARE(session.button,              MouseButton::Right);
        QCOMPARE(session.clickType,           ClickType::Double);
        QVERIFY(!session.useCurrentPosition);
        QCOMPARE(session.x,                   640);
        QCOMPARE(session.y,                   480);
        QCOMPARE(session.stopAfterClicks,     100);
        QCOMPARE(session.stopAfterSeconds,    30.0);
        QCOMPARE(session.idleWaitSeconds,     5.0);
    }

    // ── Mouse button enum coverage ────────────────────────────────────────────

    void mouseButton_left_valid()
    {
        ClickSession s(std::chrono::milliseconds(100), MouseButton::Left,
                       ClickType::Single, true, 0, 0, 0, 0.0, 0.0);
        QCOMPARE(s.button, MouseButton::Left);
    }

    void mouseButton_right_valid()
    {
        ClickSession s(std::chrono::milliseconds(100), MouseButton::Right,
                       ClickType::Single, true, 0, 0, 0, 0.0, 0.0);
        QCOMPARE(s.button, MouseButton::Right);
    }

    void mouseButton_middle_valid()
    {
        ClickSession s(std::chrono::milliseconds(100), MouseButton::Middle,
                       ClickType::Single, true, 0, 0, 0, 0.0, 0.0);
        QCOMPARE(s.button, MouseButton::Middle);
    }

    // ── Stop condition defaults ────────────────────────────────────────────────

    void stopConditions_defaultUnlimited()
    {
        ClickSession s(std::chrono::milliseconds(100), MouseButton::Left,
                       ClickType::Single, true, 0, 0, 0, 0.0, 0.0);
        QCOMPARE(s.stopAfterClicks,  0);
        QCOMPARE(s.stopAfterSeconds, 0.0);
        QCOMPARE(s.idleWaitSeconds,  0.0);
    }

    void stopConditions_nonZero_storeCorrectly()
    {
        ClickSession s(std::chrono::milliseconds(100), MouseButton::Left,
                       ClickType::Single, true, 0, 0, 50, 10.0, 3.0);
        QCOMPARE(s.stopAfterClicks,  50);
        QCOMPARE(s.stopAfterSeconds, 10.0);
        QCOMPARE(s.idleWaitSeconds,  3.0);
    }

    // ── Click type ────────────────────────────────────────────────────────────

    void clickType_single_isDefault()
    {
        ClickSession s(std::chrono::milliseconds(100), MouseButton::Left,
                       ClickType::Single, true, 0, 0, 0, 0.0, 0.0);
        QCOMPARE(s.clickType, ClickType::Single);
    }

    void clickType_double_storesCorrectly()
    {
        ClickSession s(std::chrono::milliseconds(100), MouseButton::Left,
                       ClickType::Double, true, 0, 0, 0, 0.0, 0.0);
        QCOMPARE(s.clickType, ClickType::Double);
    }
};

QTEST_MAIN(CliArgumentTests)
#include "CliArgumentTests.moc"
